### PR TITLE
Patches 1.41

### PR DESCRIPTION
**Pull request by gkapulis**
Done - IsResearchStation
Done - Module var to provide research per turn
Done - Create actual module
Done - Create chances to need research station for Stars are Planets
Done - Create notification and notification logic
Done - Ship.ResearchPerTurn Calc
Done - CanBuildResearchStations var for empire.cs
Done - Add research station condition for GoodForGoals implementation
Done - Automation window checkboxes and lists.
Done - deepspace build logic (planet)
Done - pick best researchstations
Done - Do not build defenses that are research stations
Done - build research station button - need to disallow if cannot build research stations
Done - create supply station goal logic
Done - Ship Design stats and module selection stats 
Done - do not allow AI to build reearch station with less that some ratio between storage and research point
Done - ship design issues (Low research - need cargo)
Done - universe icons
Done - Deepspace build logic star and planet
Done - Do not show research stations in colony build
Done - ResearchstationGoal
Done - refit menu
Done - if no longer researchable - remove research station and notify
Done - plan for resaerch station (researching, waiting for production, etc.)
Done - Move DistanceDisplay for planet array to public Struct
Done- Create Exostic Systems Screen List
Done - DrawResearchStaion - draw text is research station is deployed
Done - pirate missions
Done - AI simple research station build (also for player automation
Done - cache best researchstation so it could be upgraded
Done - Notificaion when a station was build
Done - refit research station  -  ability to refit all hulls 
Done - Move cargo from old research station to new research station upon refit
Done - Auto Refit logic for AI and players
Done - Defense Goals
Done - Research Planner Manager 
@gkapulis

### CODE REVIEW

File: UnitTests/Universe/ThreatMatrixTests.cs [hunk 1]
```diff
@@ -227,7 +227,7 @@ public void FindHostileClustersByDist()
             DebugVisualizeThreats(Player);
 
         ThreatCluster[] clusters = Player.Threats.FindHostileClustersByDist(pos, 55000);
-        AssertEqual(str1+str2+str3, Str(clusters));
+        AssertEqual((int)(str1+str2+str3), (int)Str(clusters));
```
Comment (RedFox20) [inline UnitTests/Universe/ThreatMatrixTests.cs:230]: Makes sense, the float values can be flaky 👍🏻

File: UnitTests/ExoticSystems/ResearchStations.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            PlanetType type = ResourceManager.Planets.RandomPlanet(PlanetCategory.GasGiant);
+            var random = new SeededRandom();
+            NewPlanet.GenerateNewFromPlanetType(random, type, scale: 1.5f, preDefinedPop: 16);
+
+            if (!NewPlanet.IsResearchable)
+                NewPlanet.SetResearchable(true, Universe.UState);
+        }
+
+        [TestMethod]
+        public void ResearchStationDeployment()
```
Comment (RedFox20) [inline UnitTests/ExoticSystems/ResearchStations.cs:36]: Could you also add a unit test for the research output rate?

File: Ship_Game/Universe/UniverseState.cs [hunk 3]
```diff
@@ -432,6 +435,26 @@ public void AddSolarSystem(SolarSystem system)
             }
         }
 
+        public void AddResearchableSolarBody(ExplorableGameObject solarBody)
+        {
+            ResearchableSolarBodies[solarBody] = new HashSet<int>();
```
Comment (RedFox20) [inline Ship_Game/Universe/UniverseState.cs:440]: This doesn't look terribly efficient. We could actually use a SmallBitSet for this, instead of a HashSet

Comment (RedFox20) [reply Ship_Game/Universe/UniverseState.cs:440]: (the performance and memory usage difference should be 50x difference)

File: Ship_Game/Universe/SolarBodies/ExoticSystemsListScreenItem.cs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            }
+        }
+
+        public override void PerformLayout()
+        {
+            int x = (int)X;
+            int y = (int)Y;
+            int w = (int)Width;
+            int h = (int)Height;
+            RemoveAll();
```
Comment (RedFox20) [inline Ship_Game/Universe/SolarBodies/ExoticSystemsListScreenItem.cs:80]: You could also use relative coordinate systems or something, so if you move one element, all the other elements will move along with it. Is the RemoveAll() strictly necessary?

File: Ship_Game/Universe/ExplorableGameObject.cs [hunk 5]
```diff
@@ -9,11 +9,37 @@ public class ExplorableGameObject : GameObject
         // this is a tiny bitset where Empire.Id is the bit index for Explored:true|false
         [StarData] SmallBitSet ExploredBy;
 
+        [StarData] public bool IsResearchable { get; private set; }
```
Comment (RedFox20) [inline Ship_Game/Universe/ExplorableGameObject.cs:12]: Lol, this single `bool` takes as much memory as the `SmallBitSet ExploredBy`. 🤡

File: Ship_Game/Universe/ExplorableGameObject.cs [hunk 6]
```diff
@@ -9,11 +9,37 @@ public class ExplorableGameObject : GameObject
...
         public void SetExploredBy(Empire empire) => ExploredBy.Set(empire.Id);
         public bool IsExploredBy(Empire empire)  => ExploredBy.IsSet(empire.Id);
+
+        public void SetResearchable(bool value, UniverseState universe)
+        {
+            IsResearchable = value;
+            if (value)
+                universe.AddResearchableSolarBody(this);
+            else
+                universe.RemoveResearchableSolarBody(this);
```
Comment (RedFox20) [inline Ship_Game/Universe/ExplorableGameObject.cs:26]:
```
            bool didChange = IsResearchable != value;
            IsResearchable = value;
            if (didChange)
            {
                if (value)
                    universe.AddResearchableSolarBody(this);
                else
                    universe.RemoveResearchableSolarBody(this);
            }
```
?

File: Ship_Game/Universe/ExplorableGameObject.cs [hunk 7]
```diff
@@ -9,11 +9,37 @@ public class ExplorableGameObject : GameObject
...
+
+            if (!empire.AI.HasGoal(g => g.IsResearchStationGoal(this)))
+                return true;
+
+            return false;
+        }
+
+        public bool IsResearchStationDeployedBy(Empire empire)
+        {
+            return IsResearchable && empire.Universe.ResearchableSolarBodies[this].Contains(empire.Id);
```
Comment (RedFox20) [inline Ship_Game/Universe/ExplorableGameObject.cs:42]: This is super inefficient, it really does need to be replaced by a SmallBitSet

File: Ship_Game/StoryAndEvents/NotificationManager.cs [hunk 8]
```diff
@@ -660,6 +716,29 @@ public void AddBoardNotification(string message, string iconPath, string action,
...
+            if (solarBody is Planet planet)
+            {
+                message = $"{planet.System.Name}: {s.Name}" +
+                    $" {Localizer.Token(GameText.ResearchStationBuiltPlanetNotify)} {planet.Name}";
+            }
+            else
+            {
+                message = $"{(solarBody as SolarSystem).Name}:" +
+                    $" {s.Name} {Localizer.Token(GameText.ResearchStationBuiltSystemNotify)}";
+            }
```
Comment (RedFox20) [inline Ship_Game/StoryAndEvents/NotificationManager.cs:731]:
```csharp
            string message;
            if (solarBody is Planet planet)
            {
                string planetNotify = Localizer.Token(GameText.ResearchStationBuiltPlanetNotify);
                message = $"{planet.System.Name}: {s.Name} {planetNotify} {planet.Name}";
            }
            else
            {
                string systemNotify = Localizer.Token(GameText.ResearchStationBuiltSystemNotify);
                message = $"{(solarBody as SolarSystem).Name}: {s.Name} {systemNotify}";
            }
```

File: Ship_Game/StoryAndEvents/NotificationManager.cs [hunk 9]
```diff
@@ -477,6 +478,39 @@ public void AddFoundSomethingInteresting(Planet p)
             }, "sd_ui_notification_encounter");
         }
 
+        public void AddReseachablePlanet(Planet p)
+        {
+            GameText text;
```
Comment (RedFox20) [inline Ship_Game/StoryAndEvents/NotificationManager.cs:483]: Use `LocalizedText` instead of GameText

File: Ship_Game/DeepSpaceBuildingWindow.cs [hunk 10]
```diff
@@ -226,7 +290,19 @@ public override void Draw(SpriteBatch batch, DrawTimes elapsed)
...
+                foreach (SolarSystem system in systems)
+                {
+                    if (system.Position.Distance(cursorWorldPos) <= system.Radius * 0.8f)
+                    {
+                        TargetSystem = system;
+                        break;
+                    }
+
+                    TargetSystem = null;
+                }
```
Comment (RedFox20) [inline Ship_Game/DeepSpaceBuildingWindow.cs:302]: This is really bad code, we already have a function in UniverseState and UniverseScreen for picking closest system to cursor world pos.

File: Ship_Game/Commands/Goals/ProcessResearchStation.cs [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+using Ship_Game.AI;
+using Ship_Game.Data.Serialization;
+using Ship_Game.Ships;
+using Ship_Game.Universe;
+using static Ship_Game.AI.ShipAI;
+
+namespace Ship_Game.Commands.Goals
+{
+    [StarDataType]
+    public class ProcessResearchStation : Goal
```
Comment (RedFox20) [inline Ship_Game/Commands/Goals/ProcessResearchStation.cs:12]: What's this goal doing? The name does not exactly explain what's going on

File: Ship_Game/AI/ExpansionAI/ResearchStationPlanner.cs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                if (solarBody.IsExploredBy(Owner)
+                    && !solarBody.IsResearchStationDeployedBy(Owner) // this bit is for performance - faster than HasGoal
+                    && !Owner.AI.HasGoal(g => g.IsResearchStationGoal(solarBody)))
+                {
+                    solarBodies.Add(solarBody);
+                }
+            }
+
+            return solarBodies.ToArray();
```
Comment (RedFox20) [inline Ship_Game/AI/ExpansionAI/ResearchStationPlanner.cs:98]: Not sure if it's a good idea to ToArray() here. This should probably be an IEnumerable or something

File: game/Content/ShipModules/Utility modules/Command modules/CIC.xml [hunk 13]
```diff
@@ -32,7 +32,7 @@
   <!-- Repair rate multiplier, see BaseSelfRepair in Globals.yaml -->
   <!-- This is 4x bigger than Bridge again, but having too insane of a modifier would break the game
        So we need to keep this reasonable at 3x-->
-  <BonusRepairRate>12.0</BonusRepairRate>
+  <BonusRepairRate>50</BonusRepairRate>
```
Comment (RedFox20) [inline game/Content/ShipModules/Utility modules/Command modules/CIC.xml:35]: Needs a fix

File: game/Content/ShipModules/Utility modules/Command modules/Bridge.xml [hunk 14]
```diff
@@ -28,7 +28,7 @@
   <PowerDraw>8</PowerDraw>
   <!-- Repair rate multiplier, see BaseSelfRepair in Globals.yaml -->
   <!-- Balanced value would be much bigger than Cockpit, to prevent cockpit exploiting -->
-  <BonusRepairRate>5.0</BonusRepairRate>
+  <BonusRepairRate>10</BonusRepairRate>
```
Comment (RedFox20) [inline game/Content/ShipModules/Utility modules/Command modules/Bridge.xml:31]: Needs a fix

File: Ship_Game/Universe/ExplorableGameObject.cs (hunk 5 above)
Comment (gkapulis) [reply Ship_Game/Universe/ExplorableGameObject.cs:12]: not sure why thats bad. we have tons of bools in the code.

File: Ship_Game/Universe/UniverseState.cs (hunk 3 above)
Comment (gkapulis) [reply Ship_Game/Universe/UniverseState.cs:440]: We are talking about around 10 objects here on average. for the whole game

**Pull request merged**